internal/auth/providers/amazon_cognito_admin.go

```diff
@@ -0,0 +1,217 @@
+					err = ErrServiceUnavailable
+				}
+			case *circuit.ErrOpenState:
+				tags = append(tags, "error:circuit_open")
+				cas.StatsdClient.Incr("provider.error", tags, 1.0)
+			default:
+				tags = append(tags, "error:invalid_response")
+				cas.StatsdClient.Incr("provider.internal_error", tags, 1.0)
+			}
+			return nil, err
+		}
+
+		tags = append(tags, fmt.Sprintf("status_code:%d", req.HTTPResponse.StatusCode))
+		cas.StatsdClient.Timing("provider.latency", time.Now().Sub(startTS), tags, 1.0)
+		cas.StatsdClient.Incr("provider.response", tags, 1.0)
+
+		for _, user := range resp.Users {
+			groupMembers = append(groupMembers, *user.Username)
+		}
+
+		if resp.NextToken == nil {
+			break
+		}
+		nextToken = *resp.NextToken
+	}
+	return groupMembers, nil
+
+}
+
+func (cas *CognitoAdminService) CheckMemberships(userName string) ([]string, error) {
+	tags := []string{
+		"provider:cognito",
+		"action:check_memberships_resource",
+	}
+	inGroups := []string{}
+
+	nextToken := ""
+	for {
+		reqParams := &cognitoidentityprovider.AdminListGroupsForUserInput{
+			Username:   &userName,
+			UserPoolId: cas.userPoolID,
+		}
+
+		startTS := time.Now()
+		req, resp := cas.adminService.AdminListGroupsForUserRequest(reqParams)
+		if nextToken != "" {
+			reqParams.SetNextToken(nextToken)
+		}
+		cas.StatsdClient.Incr("provider.request", tags, 1.0)
+
+		_, err := cas.cb.Call(func() (interface{}, error) {
+			return nil, req.Send()
+		})
+		if err != nil {
+			switch e := err.(type) {
+			case awserr.Error:
+				tags = append(tags, fmt.Sprintf("status_code:%d", req.HTTPResponse.StatusCode))
+				cas.StatsdClient.Incr("provider.response", tags, 1.0)
+				cas.StatsdClient.Incr("provider.error", tags, 1.0)
+				switch e.Code() {
+				case cognitoidentityprovider.ErrCodeTooManyRequestsException:
+					err = ErrRateLimitExceeded
+				case cognitoidentityprovider.ErrCodeInternalErrorException:
+					err = ErrServiceUnavailable
+				}
+			case *circuit.ErrOpenState:
+				tags = append(tags, "error:circuit_open")
+				cas.StatsdClient.Incr("provider.error", tags, 1.0)
+			default:
+				tags = append(tags, "error:invalid_response")
+				cas.StatsdClient.Incr("provider.internal_error", tags, 1.0)
+			}
+			return nil, err
+		}
+
+		tags = append(tags, fmt.Sprintf("status_code:%d", req.HTTPResponse.StatusCode))
+		cas.StatsdClient.Timing("provider.latency", time.Now().Sub(startTS), tags, 1.0)
+		cas.StatsdClient.Incr("provider.response", tags, 1.0)
+
+		for _, group := range resp.Groups {
+			inGroups = append(inGroups, *group.GroupName)
+		}
+
+		if resp.NextToken == nil {
+			break
+		}
+		nextToken = *resp.NextToken
+	}
+
+	return inGroups, nil
+}
+
+func (cas *CognitoAdminService) GlobalSignOut(session *sessions.SessionState) error {
+	tags := []string{
+		"provider:cognito",
+		"action:global_sign_out",
+	}
+
+	req, _ := cas.adminService.GlobalSignOutRequest(&cognitoidentityprovider.GlobalSignOutInput{
```

Comment:
Should we not check the error here?

Reply:
As far as I understand it, because we're using the `Send` method on the request (`req`), it's the `Send` call output that holds the error. The output that is ignored above contains the _response_ if `Send` returns without error. 

Brief explanation in the docs here: https://docs.aws.amazon.com/sdk-for-go/api/service/cognitoidentityprovider/#CognitoIdentityProvider.GlobalSignOutRequest

> The "output" return value will be populated with the request's response once the request completes successfully.
> Use "Send" method on the returned Request to send the API call to the service. the "output" return value is not valid until after Send returns without error.

In this case though, the sign out response should be empty: https://docs.aws.amazon.com/sdk-for-go/api/service/cognitoidentityprovider/#GlobalSignOutOutput

Reply:
Oh, this isn't an error! Its a response! Weird API.